include/irr/builtin/shaders/loaders/gltf/vertex_impl.glsl

```diff
@@ -0,0 +1,88 @@
+#include "irr/builtin/shaders/loaders/gltf/common.glsl"
+
+#ifndef _IRR_VERT_INPUTS_DEFINED_
+#define _IRR_VERT_INPUTS_DEFINED_
+
+layout (location = _IRR_POSITION_ATTRIBUTE_ID) in vec3 vPos;
+layout(location = _IRR_NORMAL_ATTRIBUTE_ID) in vec3 vNormal;
+
+#ifndef _DISABLE_COLOR_ATTRIBUTES
+layout(location = _IRR_COLOR_ATTRIBUTE_BEGINING_ID) in vec4 vColor_0;
+#ifdef _IRR_EXTRA_COLOR_ATTRIBUTE_ID_1_ENABLE
+#define _IRR_EXTRA_COLOR_ATTRIBUTE_ID_1 _IRR_COLOR_ATTRIBUTE_BEGINING_ID + 1
+layout(location = _IRR_EXTRA_COLOR_ATTRIBUTE_ID_1) in vec4 vColor_1
+#endif // _IRR_EXTRA_COLOR_ATTRIBUTE_ID_1_ENABLE
+#ifdef _IRR_EXTRA_COLOR_ATTRIBUTE_ID_2_ENABLE
+#define _IRR_EXTRA_COLOR_ATTRIBUTE_ID_2 _IRR_COLOR_ATTRIBUTE_BEGINING_ID + 2
+layout(location = _IRR_EXTRA_COLOR_ATTRIBUTE_ID_2) in vec4 vColor_2
+#endif // _IRR_EXTRA_UV_ATTRIBUTE_ID_2_ENABLE
+#endif // _DISABLE_COLOR_ATTRIBUTES
+
+#ifndef _DISABLE_UV_ATTRIBUTES
+layout (location = _IRR_UV_ATTRIBUTE_BEGINING_ID) in vec2 vUV_0;
+#ifdef _IRR_EXTRA_UV_ATTRIBUTE_ID_1_ENABLE
+#define _IRR_EXTRA_UV_ATTRIBUTE_ID_1 _IRR_UV_ATTRIBUTE_BEGINING_ID + 1
+layout (location = _IRR_EXTRA_UV_ATTRIBUTE_ID_1) in vec2 vUV_1
+#endif // _IRR_EXTRA_UV_ATTRIBUTE_ID_1_ENABLE
+#ifdef _IRR_EXTRA_UV_ATTRIBUTE_ID_2_ENABLE
+#define _IRR_EXTRA_UV_ATTRIBUTE_ID_2 _IRR_UV_ATTRIBUTE_BEGINING_ID + 2
+layout(location = _IRR_EXTRA_UV_ATTRIBUTE_ID_2) in vec2 vUV_2
+#endif // _IRR_EXTRA_UV_ATTRIBUTE_ID_2_ENABLE
+#endif // _DISABLE_UV_ATTRIBUTES
```

Comment:
glTF can really use these many attributes!?

Also check the Vulkan spec if we can declare input attributes in shader without declaring them in pipeline, **as long as the inputs are not statically used (not in the static AST tree = basically no code uses them).**
_because I know we can do this to images, etc._

Reply:
yeah it can, but I reduced it to some maxs and mins

Reply:
actually there may be more than I specified in the shader, the prefix is as follows

attribute_**INDEX**

but not for all attributes

Reply:
gotta check

Reply:
the static usage rules still wont save us from wrapping stuff

Reply:
@Crisspl what if we went "full retard" and did this
```glsl
layout(location=0u) in vec4 vPos;
layout(location=1u) in vec3 vNormal;
// uvs, colors, joint ids and weights are optional
#if NBL_GLTF_USED_VERTEX_ATTRIBUTE_COUNT>0
layout(location=2u) in vec4 vertexAttributes[NBL_GLTF_USED_VERTEX_ATTRIBUTE_COUNT];
#endif

#if NBL_GLTF_USED_TEXTURES>0
layout(set=3,binding=0u) uniform sampler2D textures[NBL_GLTF_USED_TEXTURES];
#endif

struct nbl_glsl_glTF_Parameters
{
    vec3 baseColorFactor; // could pack into a unorm 16bit together with occlusion strength
    float occlusionStrength;
    vec3 emissiveFactor; // could use RGB19E7 format to pack into a uvec2, or pack into a half float together with bumpScale
    float metallicFactor; // could pack metallic/roughness into a unorm 16bit
    float roughnessFactor;
    float bumpScale; // could be a half float
    // 5 times 4 bits for UVs, 4 bits for jointWeights, 4 bits for jointIDs, 3 bits for joint influences count
    uint uvAttributeIndices_jointWeightAttributeIndex_jointIDAttributeIndex_jointInfluences;
    // 3 times 4 bits for color attributes
    // 5 times 3 bits, baseColorTexture, metallicRoughnessTexture, derivativeMap, occlusionTexture, emissiveTexture
    uint colorAttributeIndices_materialTextureIndices;
};
layout(push_constant) PushConstants
{
    nbl_glsl_glTF_Parameters params;
} pc;
```

Then we only have [MAX_POSSIBLY_USED_ATTRIBUTES-2]x5 shader versions to compile.

Reply:
pffff, i'll inspect tomorrow, 
is it legal in all GL, GLES, VK?
how would shader introspection report this?

Reply:
lol this discussion is old af XD

Reply:
> Then we only have [MAX_POSSIBLY_USED_ATTRIBUTES-2]x5 shader versions to compile.

whye x5?

Reply:
glTF material only ever uses a maximum of 5 textures

Reply:
> 
> 
> pffff, i'll inspect tomorrow,
> is it legal in all GL, GLES, VK?
> how would shader introspection report this?

what would be illegal? and what does introspection have to do with it?

Reply:
> what would be illegal?

arrays of veretx shader inputs

> and what does introspection have to do with it?

representation of input arrays in spir-v and how spirv-cross reports this. I guess, it would report them as set of separate input points with locations in range `[array_location, array_location+array_size)`

Reply:
> arrays of veretx shader inputs
Always been legal in GLSL I think.